### PR TITLE
Add a flexible mechanism to combine user and default plugins

### PR DESCRIPTION
**Pull request by azeey**
# 🎉 New feature

Closes https://github.com/gazebosim/gz-sim/issues/796

## Summary
Implementation based on https://github.com/gazebosim/gz-sim/issues/796#issuecomment-2248821768

## TODO:
- [x] Cleanup
- [ ] Tests

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

**Comment by azeey** [comment]
@iche033 Based on your suggestion, I've simplified the various options available.
* For Server plugins, I've removed all of the options and plugins are simply appended to the list. The user can then used the `<gz:system_priority>` to configure the order the plugins.
* For GUI plugins, since the order of plugins is important but there is no mechanism similar to `<gz:system_priority>`, I've implemented the following behavior:
  * If the a plugin from the SDF also exists in the default configuration, the plugin from SDF will replace the one in the default configuration at the position/order of the plugin in the default configuration.
  * Otherwise, it's appended to the end

File: examples/worlds/camera_sensor.sdf [hunk 1]
```diff
@@ -8,127 +8,34 @@
...
-    </plugin>
 
     <scene>
       <ambient>1.0 1.0 1.0</ambient>
       <background>0.8 0.8 0.8</background>
       <grid>true</grid>
     </scene>
 
+    <gz:policies>
+      <include_default_gui_plugins>true</include_default_gui_plugins>
```
Comment (iche033) [inline examples/worlds/camera_sensor.sdf:24]:
```suggestion
      < include_gui_default_plugins >true</include_gui_default_plugins >
```

File: tutorials/gui_config.md [hunk 2]
```diff
@@ -28,10 +51,9 @@ through the command line or the SDF file.
 
 ### Default configuration
 
-Let's try this in practice. First, let's open Gazebo without passing
-any arguments:
+Let's try this in practice. First, let's open the default Gazebo world:
 
-`gz sim`
+`gz sim default.sdf`
```
Comment (iche033) [inline tutorials/gui_config.md:56]: also need to update line 82 below to be `gz sim default.sdf`

File: tutorials/server_config.md [hunk 3]
```diff
@@ -147,8 +162,84 @@ Now let's load this world:
...
+Now, let's modify the SDF file to change the policy `<include_server_config_plugins>false</include_server_config_plugins>`
+
+```xml
+<?xml version="1.0" ?>
+<sdf version="1.6">
+  <world name="fuel_preview">
+
+    <gz:policies>
+      <include_server_config_plugins>false</include_server_config_plugins>
+    <gz:policies>
```
Comment (iche033) [inline tutorials/server_config.md:182]:
```suggestion
    </gz:policies>
```

File: src/SimulationRunner.cc [hunk 4]
```diff
@@ -1586,21 +1601,60 @@ void SimulationRunner::CreateEntities(const sdf::World &_world)
...
+          [&loadedWorldPluginFileNames](const ServerConfig::PluginInfo &_pl)
+      {
+          return loadedWorldPluginFileNames.count(_pl.Plugin().Filename()) != 0;
+      };
+
+      // Remove plugin if it's already loaded so as to not duplicate world
+      // plugins.
+      defaultPlugins.remove_if(isPluginLoaded);
+
+      gzdbg << "Also loading the following default jplugins:\n";
```
[the remark this answers is not in the document]
Comment (azeey) [reply src/SimulationRunner.cc]: Fixed in https://github.com/gazebosim/gz-sim/pull/2497/commits/303e18ba127b39c75d6c9873d26a7d06f266e7f6

File: tutorials/gui_config.md [hunk 5]
```diff
@@ -16,10 +16,33 @@ There are a few places where the GUI configuration can come from:
...
+  the default configuration file. This means, if a plugin is specified in
+  both places, by default, only the one specified in the SDF file will be
+  loaded. If replacement occurs, the replacement
+  plugin will take the position of the replaced plugin in the order of plugins.
+  If replacement does not occur, the plugin is appended to the end of the list.
+
+  The main use case for this policy is for users to rely on
+  the default list of plugins and only add extra plugins they need for the
+  application. This policy is also useful for overriding the parameters of a small
+  subset of the default plugins.This is the default setting in
```
[the remark this answers is not in the document]
Comment (azeey) [reply tutorials/gui_config.md]: Fixed in https://github.com/gazebosim/gz-sim/pull/2497/commits/303e18ba127b39c75d6c9873d26a7d06f266e7f6

File: examples/worlds/camera_sensor.sdf (hunk 1 above)
Comment (azeey) [reply examples/worlds/camera_sensor.sdf]: Fixed in https://github.com/gazebosim/gz-sim/pull/2497/commits/303e18ba127b39c75d6c9873d26a7d06f266e7f6

File: tutorials/gui_config.md (hunk 2 above)
Comment (azeey) [reply tutorials/gui_config.md:56]: Updated in https://github.com/gazebosim/gz-sim/pull/2497/commits/303e18ba127b39c75d6c9873d26a7d06f266e7f6

File: tutorials/server_config.md (hunk 3 above)
Comment (azeey) [reply tutorials/server_config.md]: https://github.com/gazebosim/gz-sim/pull/2497/commits/303e18ba127b39c75d6c9873d26a7d06f266e7f6

**Review by iche033 — APPROVED** [review]
looks good. Works as expected. 

originally I was testing with separate `<gz:policies>` blocks for `include_server_config_plugins` and  `include_gui_default_plugins`, and was wondering why one of them didn't have an effect. But then realized that only one `<gz:policies>` can be present in the sdf file. Nothing needs to be done on this, just noting down for reference.

**Comment by iche033** [comment]
this test failure looks related:

```
/Users/jenkins/jenkins-agent/workspace/gz_sim-ci-pr_any-homebrew-amd64/gz-sim/src/Server_TEST.cc:514
Expected equality of these values:
  2u
    Which is: 2
  *server.SystemCount()
    Which is: 5
    ```

**Comment by iche033** [comment]
> this test failure looks related:
> 
> ```
> /Users/jenkins/jenkins-agent/workspace/gz_sim-ci-pr_any-homebrew-amd64/gz-sim/src/Server_TEST.cc:514
> Expected equality of these values:
>   2u
>     Which is: 2
>   *server.SystemCount()
>     Which is: 5
>     ```
> ```

my understanding is that an increase in no. of systems is expected due to behavior change. I increased the expected system count to fix the test, in the same way it was done to fix other Server tests. 
a0c23f2

**Comment by azeey** [comment]
> > this test failure looks related:
> > ```
> > /Users/jenkins/jenkins-agent/workspace/gz_sim-ci-pr_any-homebrew-amd64/gz-sim/src/Server_TEST.cc:514
> > Expected equality of these values:
> >   2u
> >     Which is: 2
> >   *server.SystemCount()
> >     Which is: 5
> >     ```
> > ```
> 
> my understanding is that an increase in no. of systems is expected due to behavior change. I increased the expected system count to fix the test, in the same way it was done to fix other Server tests. [a0c23f2](https://github.com/gazebosim/gz-sim/commit/a0c23f2de95152cc0503c09b2cb75a5048fb996f)

`INTEGRATION_multiple_severs` failed because we were loading the physics systems, but the test explicitly does not want it as indicated by `TestWorldSansPhysics`. My guess is that by including physics, it gets loaded twice and we run into https://github.com/gazebosim/gz-sim/issues/18

**Comment by iche033** [comment]
`gz_sim-ci-pr_any-noble-amd64` picked up many new warnings - looks like they are false positives from the new console log msgs. I'm going to merge this since all tests are passing but we should ticket an issue on this

**Comment by iche033** [comment]
> `gz_sim-ci-pr_any-noble-amd64` picked up many new warnings - looks like they are false positives from the new console log msgs. I'm going to merge this since all tests are passing but we should ticket an issue on this

#2555 https://github.com/gazebosim/gz-sim/issues/2555

**Pull request merged**